test/watch.js

```diff
@@ -0,0 +1,441 @@
+const fs = require('fs');
+const brunch = require('../lib');
+const {
+  prepareTestDir,
+  teardownTestDir,
+  fileContains,
+  fileDoesNotContains,
+  fileExists,
+  fileDoesNotExists,
+  requestBrunchServer
+} = require('./_test_helper');
+const fixturify = require('fixturify');
+
+var watcher;
+
+const watch = function(...args) {
+  watcher = brunch.watch(...args);
+};
+
+test.beforeEach.cb((t) => {
+  teardownTestDir();
+  prepareTestDir();
+
+  if (watcher) {
+    // close chokidar to prevent that it understands the fixtures being copied as new files being added
+    watcher.watcher.close();
+    if (watcher.server) {
+      watcher.server.close(t.end);
+    } else {
+      t.end();
+    }
+  } else {
+    t.end();
+  }
+});
+
+test.serial.cb('compile on file changes', t => {
+  fixturify.writeSync('.', {
+    'brunch-config.js': `module.exports = {
+      files: {
+        javascripts: {
+          joinTo: 'app.js'
+        }
+      }
+    };`,
+    'app': {
+      'assets': {
+        'index.html': '<h1>hello world</h1>'
+      },
+      'initialize.js': 'console.log("hello world")'
+    }
+  });
+
+  var calls = 0;
+
+  const onCompile = function() {
+    calls++;
+
+    if (calls === 1) {
+      fileExists(t, 'public/app.js.map');
+      fileContains(t, 'public/app.js', '//# sourceMappingURL=app.js.map');
+      fileContains(t, 'public/app.js', `require.register("initialize.js", function(exports, require, module) {
+console.log("hello world")
+});`);
+      fileContains(t, 'public/index.html', '<h1>hello world</h1>');
+
+      fs.writeFileSync('app/initialize.js', 'console.log("changed")');
+    } else if (calls === 2) {
+      fileExists(t, 'public/app.js.map');
+      fileContains(t, 'public/app.js', '//# sourceMappingURL=app.js.map');
+      fileContains(t, 'public/app.js', `require.register("initialize.js", function(exports, require, module) {
+console.log("changed")
+});`);
+      fileContains(t, 'public/index.html', '<h1>hello world</h1>');
+      t.end();
+    } else {
+      t.fail(`Unexpected number of calls ${calls}`);
+    }
+  };
+
+  watch({}, onCompile);
+});
+
+test.serial.cb('detect file addition', t => {
+  fixturify.writeSync('.', {
+    'brunch-config.js': `module.exports = {
+      files: {
+        javascripts: {
+          joinTo: 'app.js'
+        }
+      }
+    };`,
+    'app': {
+      'assets': {
+        'index.html': '<h1>hello world</h1>'
+      },
+      'initialize.js': 'console.log("hello world")'
+    }
+  });
+
+  var calls = 0;
+
+  const onCompile = function() {
+    calls++;
+
+    if (calls === 1) {
+      fileExists(t, 'public/app.js.map');
+      fileContains(t, 'public/app.js', '//# sourceMappingURL=app.js.map');
+      fileContains(t, 'public/app.js', `require.register("initialize.js", function(exports, require, module) {
+console.log("hello world")
+});`);
+      fileContains(t, 'public/index.html', '<h1>hello world</h1>');
+
+      fs.writeFileSync('app/new-file.js', 'console.log("new")');
```

Comment:
Yeah, it can indeed look somewhat like that using generators (which are supported natively by node 5 & 6, unlike async/await): https://github.com/brunch/brunch/pull/1370

```javascript
  watch({}, function *main(compilation) {
    yield compilation();
    t.true(fs.readdirSync('./node_modules').indexOf('lodash') === -1);

    const packageJson = JSON.parse(fs.readFileSync('package.json', 'utf8'));
    packageJson.dependencies.lodash = '*';
    fs.writeFileSync('package.json', JSON.stringify(packageJson, null, 2));

    yield compilation();
    t.true(fs.readdirSync('./node_modules').indexOf('lodash') !== -1);
    t.end();
  });
```